### ISSUE TITLE
Improving keyboard friendliness

### ISSUE THREAD

**Issue by qazip** [evilsocket/opensnitch#66]
Hello,

I would like to suggest that Opensnitch should be more keyboard friendly. As it is, it is possible to use the keyboard to take actions, but it's a bit cubersome to do so. So, my suggestion is the following:

1. Improve coloring/contrast of the selected button (As it is, it's hard to see if it's the "Allow" or "Deny" that's selected.

2. Move the "Take this action" dropdown to buttons; Instead of having a dropdown with "Once", "Until quit" and "Forever", I would suggest having them side by side, all visible at the same time;

3. Bind keys to all the options; For example, the numbers "1,2,3" could be shortcuts to "Once", "Until quit" and "Forever", respectively. The "w", "b", "d" and "a" could be shortcuts to "Whitelist app", "block app", "deny" and "allow", respectiviely.

Sorry if this is cherry-picking and I know there's much more important stuff to do, but I believe this could improve Opensnitch.

**Comment by Mayeu** [comment]
> Bind keys to all the options; For example, the numbers "1,2,3" could be shortcuts to "Once", "Until quit" and "Forever", respectively. The "w", "b", "d" and "a" could be shortcuts to "Whitelist app", "block app", "deni" and "allow", respectiviely.
> 
My two cents on this part, in Little Snitch I believe the shortcut for “Once”, “Until quit” and “Forever” are “o”, “u”, “f”. Which would be more user-friendly than 1, 2, 3 :)

**Comment by in1t3r** [comment]
Well I do think that falls into preference I don't like to complicate stuff so I think that 1,2,3 would be easier then searching o, u and f in the keyboard. I use i3wm so I have some of the keys that have bindings so it could be messy.

**Comment by qazip** [comment]
Well, perhaps you're right. I was thinking of actually having the numbers "1,2,3" in each button (in one of the corners, or before/after the word. anywhere really), as to be easy to see that "1" is for "Once", "2" for  "Until quit", etc.. For the Block and Deny, we could have the first letter of the word bold.

**Comment by granda** [comment]
Why not emulate the way LittleSnitch uses shortcuts?
 
![image](https://cloud.githubusercontent.com/assets/3467219/26221359/30404bea-3be4-11e7-869c-a27db23e5bb5.png)

Control to swap between "Forever" and "Until Quit" (and additional options)
Shift changes the "Until Quit" option to "Once"
Option to swap between Deny & Allow
"A" selects the radio button for "Any Connection"
"S" toggles between the three remaining buttons.
"D" opens additional information listed below:
![image](https://cloud.githubusercontent.com/assets/3467219/26221412/642031b4-3be4-11e7-9baa-8e25346e1c0f.png)

**Comment by adisbladis** [comment]
I'd say we hold off on these polishing issues until https://github.com/evilsocket/opensnitch/issues/20 is resolved and the IPC protocol and the general behaviour is nailed down.
Then we can start working on "nice to haves".

Of course anyone who wants to do the work is welcome to, but personally I'd say its not on the list of priorities right now :)

**Comment by adisbladis** [comment]
@granda I don't really like this approach.. It's too qwerty centric.

**Comment by in1t3r** [comment]
I can create different buttons for once untilquit and forever but other part is not yet built into functionality.
Also adding resolving of fqdn in prompt is easy. I agree with @adisbladis that we should work first on separation of ui and server side and then make a polished gui.

**Comment by ei8fdb** [comment]
The minimum would be to enable the UI to be navigable by keyboard only - tabbing in a logical order. This would help everyone and mean its not qwerty centric.

**Comment by phil294** [comment]
Why was this closed as completed @evilsocket? It doesn't seem like anything from the comments in this thread, including the very first one, have been solved. Can you please reopen?

With that said, I'd also love to have better keyboard navigation in the popups.

**Comment by phil294** [comment]
Hi @gustavo-iniguez-goya, can you reopen this?

More keyboard friendliness - at least, a more meaningful tab order - would be great.

**Comment by e3dio** [comment]
> Move the dropdown to buttons

I agree the dropdown selections involve too much clicking around on stuff. If I want to change duration and choose action I have to click dropdown, click duration, click dropdown, click action, lots of effort involved. If all options are displayed as checkboxes or buttons, that is a single click or key press and no searching around

**Comment by gustavo-iniguez-goya** [comment]
Hi @phil294 ,

There have been some improvements since this issue was opened. You can select popup buttons with TAB or SHIFT+TAB. 
The order is:

![Captura de pantalla de 2025-01-06 18-19-40](https://github.com/user-attachments/assets/8597c6bc-9ef4-49ed-993c-9760253ea2c1)


> Improve coloring/contrast of the selected button (As it is, it's hard to see if it's the "Allow" or "Deny" that's selected.

The default themes have a hint to indicate that a widget is selected:

![image](https://github.com/user-attachments/assets/943d78cd-85e9-4be2-8ed9-8a28712a832a)

And on the other hand, the button which is active is the one with the countdown.

Right now the [+] button is selected by default, but it could be the default action (allow, deny).

Regarding the keyboard shortcuts, it's just a matter of defining some defaults, and allow the users redefine them from the Preferences dialog.

**Comment by phil294** [comment]
I'm sorry, you're right! I was on an outdated version by accident back then. Thanks! :)

**Comment by phil294** [comment]
Hi again @gustavo-iniguez-goya, just re-read your comment and saw the final sentence:

> Regarding the keyboard shortcuts, it's just a matter of defining some defaults, and allow the users redefine them from the Preferences dialog.

True, but since this doesn't exist yet: Can you please reopen this issue?

Also, I wanted to have a go at this myself, but I'm on Arch Linux and can't/don't want to use pip, and system-packaged python packages appear to be incompatible (#491) so this is kind of a blocker. But having an open issue for this would make sense.

**Comment by gustavo-iniguez-goya** [comment]
reopened @phil294
